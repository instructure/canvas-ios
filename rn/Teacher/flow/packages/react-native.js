//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
