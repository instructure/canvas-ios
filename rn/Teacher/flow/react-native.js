//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
