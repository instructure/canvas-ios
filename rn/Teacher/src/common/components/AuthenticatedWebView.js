//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React from 'react'
import ActivityIndicatorView from './ActivityIndicatorView'
import canvas from '../../canvas-api'
import CanvasWebView from './CanvasWebView'

export default class AuthenticatedWebView extends React.Component<any, any> {
  webView: CanvasWebView

  constructor (props: any) {
    super(props)
    this.state = {
      loading: true,
    }
  }

  injectJavaScript = (script: string) => {
    return this.webView.evaluateJavaScript(script)
  }

  captureRef = (c: CanvasWebView) => {
    this.webView = c
  }

  componentDidMount () {
    if (this.props.source && this.props.source.uri) {
      this.update()
    } else {
      this.setState({
        loading: false,
      })
    }
  }

  async update () {
    this.setState({
      loading: true,
    })
    const uri = this.props.source.uri
    let authedUri = uri
    try {
      const result = await canvas.getAuthenticatedSessionURL(uri)
      if (result.data.session_url) {
        authedUri = result.data.session_url
      }
    } catch (e) {}
    this.setState({
      uri: authedUri,
      loading: false,
    })
  }

  render () {
    if (this.state.loading) {
      return <ActivityIndicatorView />
    }

    const uri = this.state.uri
    const props = {
      ...this.props,
      source: uri ? { uri } : this.props.source,
    }
    return <CanvasWebView {...props} ref={this.captureRef} />
  }
}
