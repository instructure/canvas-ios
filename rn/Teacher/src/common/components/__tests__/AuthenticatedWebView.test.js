//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import 'react-native'
import React from 'react'
import AuthenticatedWebView from '../AuthenticatedWebView'

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

jest.unmock('../AuthenticatedWebView.js')

test('AuthenticatedWebView renders', async () => {
  jest.useFakeTimers()
  let tree = renderer.create(
    <AuthenticatedWebView source={{ uri: 'http://fillmurray.com/100/100' }}/>
  )
  await tree.getInstance().update()
  jest.runAllTicks()
  expect(tree.toJSON()).toMatchSnapshot()
})

test('AuthenticatedWebView renders without uri', async () => {
  let tree = renderer.create(
    <AuthenticatedWebView source={{ html: 'html is the coolest' }}/>
  )
  expect(tree.toJSON()).toMatchSnapshot()
})

test('AuthenticatedWebView can inject javascript and not explode', () => {
  let tree = renderer.create(
    <AuthenticatedWebView source={{ uri: 'http://fillmurray.com/100/100' }}/>
  )
  const instance = tree.getInstance()
  instance.webView = {
    evaluateJavaScript: jest.fn(),
  }
  tree.getInstance().injectJavaScript(`console.log('hello')`)
  expect(tree.getInstance().webView.evaluateJavaScript).toHaveBeenCalled()
})
