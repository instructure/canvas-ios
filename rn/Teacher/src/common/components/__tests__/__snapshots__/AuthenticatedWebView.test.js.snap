// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthenticatedWebView renders 1`] = `
<View>
  <CoreWebView
    bounces={true}
    onHeightChange={[Function]}
    onNavigation={[Function]}
    source={
      Object {
        "uri": "http://fillmurray.com/100/100",
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`AuthenticatedWebView renders without uri 1`] = `
<View>
  <CoreWebView
    bounces={true}
    onHeightChange={[Function]}
    onNavigation={[Function]}
    source={
      Object {
        "html": "html is the coolest",
      }
    }
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
